src/wast-parser.cc

```diff
@@ -1488,6 +1497,96 @@ Result WastParser::ParsePlainInstr(std::unique_ptr<Expr>* out_expr) {
   return Result::Ok;
 }
 
+// Simd
+// The expected V128 const lists is:
+// i32 0xXXXXXXXX 0xXXXXXXXX 0xXXXXXXXX 0xXXXXXXXX
+Result WastParser::ParseSimdConstList(ConstVector* consts) {
+  WABT_TRACE(ParseSimdConstList);
+  Literal literal;
+
+  // Check input const type.
+  if (!PeekMatch(TokenType::ValueType))
+    return Result::Error;
+
+  Type in_type = Consume().type();
+
+  switch (in_type) {
+    case Type::I32: {
+      // Meet expected "i32" token. start parse 4 i32 consts
+      for(int i=0; i<4; i++) {
+        Const const_;
+        const_.loc = GetLocation();
+
+        // Expected one 0xXXXXXXXX number
+        if (!PeekMatch(TokenType::Nat))
+          return ErrorExpected({"an Nat literal"}, "123");
+
+        Literal literal = Consume().literal();
+
+        string_view sv = literal.text;
+        const char* s = sv.begin();
+        const char* end = sv.end();
+        Result result;
+
+        const_.type = Type::I32;
+
+        result =
+            ParseInt32(s, end, &(const_.u32), ParseIntType::SignedAndUnsigned);
+
+        if (Failed(result)) {
+          Error(const_.loc, "invalid literal \"%s\"", literal.text.c_str());
+        }
+
+        consts->push_back(const_);
+      }
+      break;
+    }
+
+    default:
+      assert(!"ParseSimdConstListcalled with invalid input const type");
+      return Result::Error;
+  }
+
+  return Result::Ok;
+}
+
+Result WastParser::ParseSimdConst(Const* const_) {
+  WABT_TRACE(ParseSimdConst);
+  Opcode opcode = Consume().opcode();
+
+  const_->loc = GetLocation();
+
+  ConstVector simdconsts;
+  CHECK_RESULT(ParseSimdConstList(&simdconsts));
+
+  switch (opcode) {
+    case Opcode::V128Const: {
+      const_->type = Type::V128;
+      // 1. check the const type and range.
+      int nSize = simdconsts.size();
+      assert(nSize == 4);
+      // 2. Copy it into const_.v128_bits.
+      union i128bits{
+        int32_t i32Array[4];
+        uint8_t i8Array[16];
```

Comment:
Why is the union needed? I think you can use the i32array directly in memcpy below.